src/deploy/functions/backend.ts

```diff
@@ -0,0 +1,356 @@
+// to short circuit early (not like there's much point in optimizing JS...)
+export const sameFunctionName = (func: FunctionSpec) => (test: FunctionSpec) => {
+  return func.id === test.id && func.region === test.region && func.project == test.project;
+};
+
+export function scheduleName(schedule: ScheduleSpec, appEngineLocation: string) {
+  return `projects/${schedule.project}/locations/${appEngineLocation}/jobs/${schedule.id}`;
+}
+
+// This method uses a separate struct that PubSubSpec conforms to so that a schedule
+// can be passed as well
+export function topicName(topic: { project: string; id: string }) {
+  return `projects/${topic.project}/topics/${topic.id}`;
+}
+
+/*
+ ** The naming pattern used to create a Pub/Sub Topic or Scheduler Job ID for a given scheduled function.
+ ** This pattern is hard-coded and assumed throughout tooling, both in the Firebase Console and in the CLI.
+ ** For e.g., we automatically assume a schedule and topic with this name exists when we list funcitons and
+ ** see a label that it has an attached schedule. This saves us from making extra API calls.
+ ** DANGER: We use the pattern defined here to deploy and delete schedules,
+ ** and to display scheduled functions in the Firebase console
+ ** If you change this pattern, Firebase console will stop displaying schedule descriptions
+ ** and schedules created under the old pattern will no longer be cleaned up correctly
+ */
+export function scheduleIdForScheduledFunction(cloudFunction: FunctionSpec) {
+  return `firebase-schedule-${cloudFunction.id}-${cloudFunction.region}`;
+}
+
+export function toGCFv1Function(
+  cloudFunction: FunctionSpec,
+  sourceUploadUrl: string
+): Omit<gcf.CloudFunction, gcf.OutputOnlyFields> {
+  if (cloudFunction.apiVersion != 1) {
+    throw new FirebaseError(
+      "Trying to create CloudFunction with wrong API. This should never happen"
+    );
+  }
+
+  const gcfFunction: Omit<gcf.CloudFunction, gcf.OutputOnlyFields> = {
+    name: functionName(cloudFunction),
+    sourceUploadUrl: sourceUploadUrl,
+    entryPoint: cloudFunction.entryPoint,
+    runtime: cloudFunction.runtime,
+  };
+
+  if (isEventTrigger(cloudFunction.trigger)) {
+    gcfFunction.eventTrigger = {
+      eventType: cloudFunction.trigger.eventType,
+      resource: cloudFunction.trigger.eventFilters.resource,
+      // Service is unnecessary and deprecated
+    };
+    if (cloudFunction.trigger.retry) {
+      gcfFunction.eventTrigger.failurePolicy = { retry: {} };
+    }
+  } else {
+    gcfFunction.httpsTrigger = {
+      securityLevel: cloudFunction.trigger.httpsOnly ? "SECURE_ALWAYS" : "SECURE_OPTIONAL",
+    };
+  }
+
+  proto.renameIfPresent(
+    gcfFunction,
+    cloudFunction,
+    "timeout",
+    "timeoutSeconds",
+    proto.durationFromSeconds
+  );
+  proto.renameIfPresent(gcfFunction, cloudFunction, "serviceAccountEmail", "serviceAccount");
+  proto.copyIfPresent(gcfFunction, cloudFunction, "availableMemoryMb");
```

Comment:
Can `copyIfPresent` take a variadic final arg so it doesn't have to be repeated for each field?

Reply:
Now I remember why I didn't do this. I had an optional transform function. Since it was only used once I made that one case an explicit copy.